tls/test_tls_alpn_basic.py

```diff
@@ -0,0 +1,176 @@
+            }
+        )
+
+    @parameterize.expand(
+        [
+            param(
+                name="order_https",
+                tempesta_proto="https",
+                protocols=["http/1.1", "h2"],
+            ),
+            param(
+                name="order_h2",
+                tempesta_proto="h2",
+                protocols=["h2", "http/1.1"],
+            ),
+            param(
+                name="mixed_https",
+                tempesta_proto="https,h2",
+                protocols=["http/1.1"],
+            ),
+            param(
+                name="mixed_h2",
+                tempesta_proto="https,h2",
+                protocols=["h2"],
+            ),
+        ]
+    )
+    @dmesg.unlimited_rate_on_tempesta_node
+    def test_proto(self, name, tempesta_proto, protocols):
+        self.init_config(tempesta_proto)
+        self.start_tempesta()
+
+        self.assertTrue(ssl.HAS_ALPN, "`ssl` library backend does not support ALPN")
+
+        hostname = tf_cfg.cfg.get("Tempesta", "hostname")
+        context = ssl.create_default_context()
+        context.set_alpn_protocols(protocols)
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        with socket.create_connection((hostname, 443)) as sock:
+            with context.wrap_socket(sock, server_hostname="tempesta-tech.com") as ssock:
+                self.assertEqual(
+                    ssock.selected_alpn_protocol(),
+                    protocols[0],
+                    "wrong protocol has been prioritized",
+                )
+
+    @parameterize.expand(
+        [
+            param(
+                name="https",
+                tempesta_proto="https",
+                protocols=["h2"],
+                msg="ClientHello: cannot find matching ALPN for h2",
+            ),
+            param(
+                name="h2",
+                tempesta_proto="h2",
+                protocols=["http/1.1"],
+                msg="ClientHello: cannot find matching ALPN for http/1.1",
+            ),
+        ]
+    )
+    @dmesg.unlimited_rate_on_tempesta_node
+    def test_negative(self, name, tempesta_proto, protocols, msg):
+        """
+        Try to establish handshake with unsupported protocol in ALPN protocols list. Tempesta
+        must not finish handshake successfully.
+        """
+        self.init_config(tempesta_proto)
+        self.start_tempesta()
+
+        self.assertTrue(ssl.HAS_ALPN, "`ssl` library backend does not support ALPN")
+
+        hostname = tf_cfg.cfg.get("Tempesta", "hostname")
+        context = ssl.create_default_context()
+        context.set_alpn_protocols(protocols)
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        with socket.create_connection((hostname, 443)) as sock:
+            try:
+                ssock = context.wrap_socket(sock, server_hostname="tempesta-tech.com")
+            except:
```

Comment:
Please add the expected exception here. Or you can use `with self.assertRaises` construction here.